Setup ability to provide authenticated ingresses to arbitrary web apps

yuvipanda opened the pull request:
- Setup dex to proivde oidc authentication

  JupyterHub provides oauth2, but *not* oidc. This sucks, but
  as with most problems in computer science, we can solve this
  with another layer of indirection.

  In this case, we use https://dexidp.io to setup an oidc provider
  that is itself authenticated via JupyterHub. This is transparent
  to the user (because we don't require any consent screens), so in
  practice it's gonna look like JupyterHub is providing oidc itself.

  Instead of automatically injecting config from the deployer,
  we're instead manually copy-pasting config into the .values
  files directly. This helps with right to replicate, and keeps our
  deployer smaller

- Setup oauth2-proxy alongside dex

  We'll use nginx-ingress to authenticate against our hub, and we'll
  run an oauth2-proxy as a sidecar with dex so it can be used to
  authenticate requests!

- Add staticSites that can be protected by oauth2-proxy
  
    - Uses nginx to serve these, and a 5min git pull to pull a git
      repo. Copied from earlier docs_service code
    - Use ingress provider to specify the path under which the
      static site is available, rather than the hub proxy. This is
      simpler and allows us to put them anywehere (like
      staging.2i2c.cloud/textbook rather than
      staging.2i2c.cloud/services/textbook)
    - Use nginx-ingress to provide authentication for the static
      server, via https://kubernetes.github.io/ingress-nginx/examples/auth/oauth-external-auth/

-  Add support for pulling from private github repos
    
    Instead of using deploy keys, we use
    https://github.com/yuvipanda/git-credential-helpers,
    which is also used for pulling from private repos with nbgitpuller.
    
    I've created a new GitHub app for the 2i2c staging hub,
    at https://github.com/apps/2i2c-staging-hub-static-pull and 'installed'
    it for the private repo at https://github.com/yuvipanda/test-repo-push.
    This allows the static-sites pod to pull it, and it's available at
    https://staging.2i2c.cloud/textbook/ now
- Remove `docs-service`
# TODO

- [x] Add documentation

Fixes #1500

damianavila commented:
cc @GeorgianaElena for awareness since this one will be needed for the upcoming #1300.

yuvipanda commented:
You can test this out at staging.2i2c.cloud/textbook!

GeorgianaElena commented on config/clusters/2i2c/staging.values.yaml:
```diff
@@ -23,6 +30,13 @@ jupyterhub:
           name: 2i2c
           url: https://2i2c.org
   hub:
+    services:
+      dex:
```
We don't need this service to be visible from the `services` menu right?
What do you think about using `display: false` to hide it (the same we're doing for dask-gateway)?

I think it might be confusing for users to have services throwing 404s

yuvipanda replied:
Good call, I've updated it @GeorgianaElena!

GeorgianaElena commented on config/clusters/2i2c/staging.values.yaml:
```diff
@@ -23,6 +30,13 @@ jupyterhub:
+        url: http://dex:5556
+        oauth_redirect_uri: https://staging.2i2c.cloud/services/dex/callback
+        oauth_no_confirm: true
+      oauth2-proxy:
+        url: http://dex:9000
```
Same:
```suggestion
      oauth2-proxy:
        url: http://dex:9000
        display: false
```

choldgraf commented:
This looks great to me - I created a second issue (linked in the parent issue) to track documenting this as well

yuvipanda commented:
@choldgraf i've added docs on how to enable it and how frequently it gets updated. Not user facing though.

yuvipanda commented:
I've added support for pulling from private repos, and also added docs on how to do that! 

I'm going to mark this as ready to review. Need to write a topic/ help page on how this actually works though, but doesn't need to block this PR

consideRatio commented on config/clusters/2i2c/staging.values.yaml:
```diff
@@ -1,9 +1,18 @@
+  enabled: true
+  hubHostName: staging.2i2c.cloud
+
+staticSites:
+  enabled: true
+  repo: https://github.com/yuvipanda/test-repo-push
+  branch: master
+  host: staging.2i2c.cloud
+  path: /textbook
+  githubApp:
+    enabled: true
```
I think an increased verbosity in the naming of these will be helpful. If you read the config names and feel very confident what they are about.

- `staticSites`
   For now I understand there is an assumption to only support exposing one static content website. That seems fine to me and support for more can be added over time with some refactoring. The name is currently `staticSites` in plural. What about `staticWebsite` instead? I think that helps readers of this config better understand what this is about, because `site` is a little vague in the context of reading config assuming you try to understand what what the config is about.
   
- `repo` and `branch` are related to a git repository to be synced. `host` and `path` is about Ingress resources and how to access them. `githubApp` was something I didn't understand what it was about until I read about it, but I now understand it as configuration of credentials to access the git repo to sync from.

  I propose more explicit names like `gitRepoUrl`, `gitRepoBranch`, `ingressHost`, `ingressPath`, and perhaps `gitRepoCredentials.enabled`, with nested values `githubAppId`, `githubAppKey`.
  
I've grown to value clearly named config as even more important than clearly named variables and functions, because config names are like variable names exposed to an even wider audience.

yuvipanda replied:
@consideRatio thank you for this very helpful feedback. I've renamed and moved around config so it is more self descriptive. I have preferred nesting over compound words, as I think that expresses intent clearer in a hierarchical config format like YAML. I hope this is good from your pov.

yuvipanda replied:
Thanks a lot for this thoughtful response, @consideRatio. You've convinced me easily it should be called `ingress` and not `url` and I've made that change.

Re nested vs flat, I very much prefer nested for the following reasons:

1. Avoids repetition in the name, making it easier to read the config.
2. Allows expansion in the future - we can add a `source.type` that can be set to `git` or `google-drive` or whatever, and we can add another `googleDrive` key at that point without breaking anything under `git`. This combination of setting `type` and then nesting keys under that is one of my fav k8s patterns.

Thank you for thinking deeply about this, and communicating it clearly!

consideRatio commented on helm-charts/basehub/templates/dex/configmap.yaml:
```diff
@@ -0,0 +1,67 @@
+      skipApprovalScreen: true
+
+    connectors:
+    - type: oauth
+      id: hub
+      name: hub
+      config:
+        clientID: service-dex
+        # Env vars are expanded via gomplate, which is present in the
+        # upstream dex docker image
+        clientSecret:  {{ "{{" }} .Env.HUB_OAUTH2_CLIENT_SECRET {{ "}}" }}
+        redirectURI: https://{{ .Values.dex.hubHostName }}/services/dex/callback
+        userIDKey: name
+        tokenURL: http://proxy-public/hub/api/oauth2/token
+        authorizationURL: https://{{ .Values.dex.hubHostName }}/hub/api/oauth2/authorize
+        userInfoURL: http://proxy-public/hub/api/user
+
+    staticClients:
+    - id: oauth2-proxy
+      redirectURIs:
+      -  https://{{ .Values.dex.hubHostName }}/services/oauth2-proxy/oauth2/callback
+      name: oauth2-proxy
+      # Env vars are expanded via gomplate, which is present in the
+      # upstream dex docker image
+      secret: {{ "{{" }} .Env.OAUTH2_PROXY_CLIENT_SECRET {{ "}}" }}
```
Breaks common YAML practice of indententing list entries, valid though so just formatting aspect.

consideRatio commented on helm-charts/basehub/templates/dex/deployment.yaml:
```diff
@@ -0,0 +1,82 @@
+            mountPath: /srv/db
+          # Needs to be args, not cmd - this allows gomplate based
+          # expansion of config file
+          args:
+          - dex
+          - serve
+          - /srv/config/dex.yaml
+        - name: oauth2-proxy
+          image: quay.io/oauth2-proxy/oauth2-proxy:v7.3.0
+          command:
+          - oauth2-proxy
+          - --config=/srv/config/oauth2-proxy.cfg
+          volumeMounts:
+          - name: config
+            mountPath: /srv/config
```
Indentation of list entries, formatting nitpicking.

consideRatio commented on helm-charts/basehub/templates/dex/deployment.yaml:
```diff
@@ -0,0 +1,82 @@
+          ports:
+            - name: oauth2-proxy
+              containerPort: 9000
+          env:
+          # This is read by oauth2-proxy
+          - name: OAUTH2_PROXY_COOKIE_SECRET
+            valueFrom:
+              secretKeyRef:
+                name: dex
+                key: oauth2Proxy.cookieSecret
+          - name: OAUTH2_PROXY_CLIENT_SECRET
+            valueFrom:
+              secretKeyRef:
+                name: dex
+                key: oauth2Proxy.clientSecret
```
Indentation nitpicking again.

yuvipanda replied:
I personally prefer not indenting lists, but I also want to make these conversations go away asap so i'll just indent it :) Thank you for helping us keep things consistent, @consideRatio. Ideally an autoformatter would do this but I think those don't work on go template files like this.

yuvipanda replied:
I re-read my comment and I have no idea if i prefer indenting lists or not indenting lists lol. I've just followed your suggestions now.

consideRatio commented on helm-charts/basehub/templates/dex/secret.yaml:
```diff
@@ -0,0 +1,21 @@
+stringData:
+  {{- $k8sState := lookup "v1" "Secret" .Release.Namespace "dex" | default (dict "data" (dict)) }}
+
+  {{- if hasKey $k8sState.data "oauth2Proxy.clientSecret" }}
+    oauth2Proxy.clientSecret: {{ index $k8sState.data "oauth2Proxy.clientSecret" | b64dec }}
+  {{- else }}
+    oauth2Proxy.clientSecret: {{ include "randHex" 64  }}
+  {{- end }}
+
+  {{- if hasKey $k8sState.data "oauth2Proxy.cookieSecret" }}
+    oauth2Proxy.cookieSecret: {{ index $k8sState.data "oauth2Proxy.cookieSecret" | b64dec }}
+  {{- else }}
+    oauth2Proxy.cookieSecret: {{ include "randHex" 16 }}
+  {{- end }}
```
It may read clearer in this case, but it can also encourage this kind of practice to indent content relative to logic in helm templates.

+1 for consistent use of indentation practice, in this case it would mean to dedent line 11, 13, and 19 two spaces.

yuvipanda replied:
Done

consideRatio commented on helm-charts/basehub/templates/static/deployment.yaml:
```diff
@@ -0,0 +1,97 @@
+            name: static-sites
+        - name: content
+          emptyDir: {}
+        {{ if .Values.staticSites.githubApp.enabled }}
+        - name: git-config
+          secret:
+            secretName: static-sites
+        {{ end }}
```
```suggestion
        {{- if .Values.staticSites.githubApp.enabled }}
        - name: git-config
          secret:
            secretName: static-sites
        {{- end }}
```

consideRatio commented on helm-charts/basehub/templates/static/deployment.yaml:
```diff
@@ -0,0 +1,97 @@
+          volumeMounts:
+            - name: content
+              mountPath: /srv/content
+            {{ if .Values.staticSites.githubApp.enabled }}
+            - name: git-config
+              mountPath: /etc/gitconfig
+              subPath: gitconfig
+              readOnly: true
+            - name: git-config
+              mountPath: /etc/github/github-app-private-key.pem
+              subPath: github-app-private-key.pem
+              readOnly: true
+            {{ end }}
```
```suggestion
            {{- if .Values.staticSites.githubApp.enabled }}
            - name: git-config
              mountPath: /etc/gitconfig
              subPath: gitconfig
              readOnly: true
            - name: git-config
              mountPath: /etc/github/github-app-private-key.pem
              subPath: github-app-private-key.pem
              readOnly: true
            {{- end }}
```

consideRatio commented on helm-charts/basehub/templates/static/deployment.yaml:
```diff
@@ -0,0 +1,97 @@
+          volumeMounts:
+            - name: content
+              mountPath: /srv/content
+            {{ if .Values.staticSites.githubApp.enabled }}
+            - name: git-config
+              mountPath: /etc/gitconfig
+              subPath: gitconfig
+              readOnly: true
+            - name: git-config
+              mountPath: /etc/github/github-app-private-key.pem
+              subPath: github-app-private-key.pem
+              readOnly: true
+            {{ end }}
```
```suggestion
            {{- if .Values.staticSites.githubApp.enabled }}
            - name: git-config
              mountPath: /etc/gitconfig
              subPath: gitconfig
              readOnly: true
            - name: git-config
              mountPath: /etc/github/github-app-private-key.pem
              subPath: github-app-private-key.pem
              readOnly: true
            {{- end }}
```

consideRatio commented on helm-charts/basehub/templates/static/deployment.yaml:
```diff
@@ -0,0 +1,97 @@
+            - name: nginx
+              containerPort: 8080
+          volumeMounts:
+          - name: config
+            mountPath: /etc/nginx/conf.d/default.conf
+            subPath: nginx.conf
+          - name: content
+            mountPath: /srv/content
```
```suggestion
            - name: config
              mountPath: /etc/nginx/conf.d/default.conf
              subPath: nginx.conf
            - name: content
              mountPath: /srv/content
```

consideRatio commented on helm-charts/basehub/templates/static/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+            name: static-sites
+            port:
+              number: 80
+{{- end }}
```
```suggestion
{{- end }}

```

consideRatio commented on helm-charts/basehub/templates/dex/pvc.yaml:
```diff
@@ -0,0 +1,12 @@
+  resources:
+    requests:
+      storage: 1Gi
+{{- end }}
```
```suggestion
{{- end }}

```

consideRatio commented on helm-charts/basehub/templates/static/ingress.yaml:
```diff
@@ -0,0 +1,24 @@
+spec:
+  ingressClassName: nginx
+  rules:
+  - host: {{ .Values.staticSites.host}}
```
```suggestion
  - host: {{ .Values.staticSites.host }}
```

consideRatio reviewed (commented):
Nice! I nitpicked about YAML formatting details and reflected on more verbose naming of the config to help whoever ends up reading this config more confidently guessing what this is about just based on the config names.

Great feature and I think the implementation seems solid overall!

yuvipanda commented:
@consideRatio I appreciate the thought you put into keeping things consistent, thank you. I've addressed all your concerns I think.

consideRatio reviewed (approved):
Thank you @yuvipanda!!!

GeorgianaElena reviewed (approved):
LGTM 🚀

GeorgianaElena commented:
> Need to write a topic/ help page on how this actually works though, but doesn't need to block this PR

Should we open an issue to track this?

GeorgianaElena commented on docs/howto/features/static-sites.md:
```diff
@@ -0,0 +1,162 @@
+4. Under 'Repository permissions', select 'Read' for 'Contents'.
+
+5. Under 'Where can this GitHub App be installed?', select 'Any account'. This will
+   enable users to push to their own user repositories or other organization repositaries,
```
```suggestion
   enable users to push to their own user repositories or other organization repositories,
```
